"a la carte" import + traits

eed3si9n opened the pull request:
Implemented "a la carte" import.
1. Ops are moved out of pickling package object, and will be part of `import all._`.
2. `allPicklers` are broken down into smaller traits so third-party protocols can lock down to specific data structures they want to support.
3. Added `generate` method to `SPickler` and `Unpickler` companion objects.
### A la carte usage

``` scala
import scala.pickling._ // This imports names only
import json._           // Imports PickleFormat
import static._         // Avoid runtime pickler
import ops._            // Inject pickle and unpickle methods
// Alternatively import pickle function
// import scala.pickling.functions._ 

// Import picklers for specific types
import allPicklers.{ stringPickler, intPickler, refUnpickler, nullPickler }

case class Pumpkin(kind: String)
// Manually generate a pickler using macro
implicit val pumpkinPickler = SPickler.generate[Pumpkin]
implicit val pumpkinUnpickler = Unpickler.generate[Pumpkin]

val pckl = Pumpkin("Kabocha").pickle
val pump = pckl.unpickle[Pumpkin]
```
### Basic usage

Breaking change is switch to `scala.pickling.Defaults._` instead of `allPicklers._`.

``` scala
import scala.pickling._, scala.pickling.Defaults._, json._

val pckl = List(1, 2, 3, 4).pickle
val lst = pckl.unpickle[List[Int]]
```

eed3si9n committed:
Split enrich-my-library ops out of package object

As the first step of “a la carte”, pickling package no longer
introduces ops as implicit classes.
This allows third-party to reuse pickling facility for serialization
with their own ops.

eed3si9n committed:
Split up CorePicklersUnpicklers into smaller traits

havocp commented on core/src/main/scala/pickling/pickler/AllPicklerUnpicklers.scala:
```diff
@@ -0,0 +1,31 @@
+  with LinearSeqPicklers
+  with IndexedSeqPicklers
+  with SeqPicklers
+  with IterablePicklers
```
should these be grouped somehow, CollectionPicklers maybe?

eed3si9n replied:
Done

havocp commented:
Looks nice!

eed3si9n committed:
Split LowPriorityPicklers into CollectionPicklers and AnyUnpicklers

jsuereth commented on core/src/main/scala/pickling/package.scala:
```diff
@@ -1,18 +1,36 @@
+    new AllPicklers with Ops {}
+
+  trait Ops extends ToPickleOps with ToUnpickleOps {}
+  trait ToPickleOps {
```
These really do add a TON of bytecode to the project.

jsuereth commented on core/src/main/scala/pickling/pickler/Array.scala:
```diff
@@ -0,0 +1,12 @@
+import scala.pickling._
+import scala.collection.generic.CanBuildFrom
+
+trait ArrayPicklers {
```
All of these "ala carte" traits need to be documented in terms of what they can handle.

Additionally, it'd be ideal for those which could be statically generated and kept somewhere to be initiated in *one* place rather than regenerated via the implicit macros.   For array you should document that a compile-time-optimisation is to do that.

jsuereth commented on core/src/main/scala/pickling/pickler/List.scala:
```diff
@@ -0,0 +1,134 @@
+import scala.language.experimental.macros
+import scala.language.higherKinds
+
+trait ListPicklerUnpicklerMacro extends CollectionPicklerUnpicklerMacro {
```
This is for the macro implementations and doesn't belong in the pickler package, IMHO.   It's not really an "ala carte" thing.

jsuereth committed:
Fixes from discussion.

eed3si9n committed:
Use package scala.pickling, package pickler

phaller commented:
@eed3si9n: let me know when you'd like this to be reviewed for merge. Thanks!

eed3si9n committed:
scala.pickling.all -> scala.pickling.Defaults

eed3si9n committed:
import ops._ within macros

- Move functions out to PickleFunctions trait.
- import ops._ from macros since macro implementations use ops.
- Add test to check you can generate pickler without having opts in the
call site scope.

eed3si9n committed:
Turn datePickler back to val

eed3si9n committed:
Remove unnecessary objects

eed3si9n commented:
@phaller @jsuereth Implemented all changes we discussed.

1. Code organized with nested `package scala.pickling` and `package pickler`.
2. `scala.pickling.all` is now `scala.pickling.Defaults`.
3. Macro implementations import `ops._` from within.
4. `datePickler` is val.
5. Removes unnecessary objects extending a la carte traits.

jsuereth commented:
The A.la.carte pattern is still missing documentation via Scaladoc w/in the project on intended usage.

My main complaint is about the PickleFunctions trait.   I'm highly dubious that users of this framework will want to not use a static method (or delegate to one) for pickling, while the rest of the "cake bake-off" around default pickler implementations *or* whether or not to "enhance" `Any` with `.pickle`/`.unpickle`, I'm really unable to swallow a non-static `pickle(obj)(format, pickler)` method.   Can you provide me any example where having these static methods would be less desirable than in a trait?   Can you justify the additional bytecode (and strain on the JIT cache) for this?   Even for our own usage, if we provide our own methods of pickling/unpickling there's no way the SIGNATURE would look the same, so I can't imagine wanting to "bake" something w/ PickleFunctions.

My secondary complaint is around self types.   They aren't providing any real value IMHO.   You're not avoiding any override/initialization issues, you still force users to have extension *and* they still need to use early-initializers if they run into cake-baking issues.   We had discussed this one the phone, but wanted to see why you migrated back to self-types.

eed3si9n commented:
In my opinion, an internally consistent API is easier to learn, and easier to use and/or extend.

There are three types of audiences for Pickling:
1. Users of `Defaults` mode.
2. Some someone who wants to use Pickling in a limited scope like [static-only-with-manual-pickler.scala](https://github.com/eed3si9n/pickling/blob/123c5580724cd4c9718a8ff27515c84e6f93d1c6/core/src/test/scala/pickling/run/static-only-with-manual-pickler.scala).
3. The library authors who wants to extend (and limit) Pickling.

A la carte at the value level is for 2., and a la carte at the trait level is for 3.
I think functions, typeclass instance, and functions should all be provided in consistent fashion. Having a trait allows someone to make a trait stack like:

```scala
package foo

import scala.pickling._
package object serialization extends PickleFunctions with picklers.AllPicklers {}
```

I still think Ops should be trait for this reason, and we saw that the bytecode size wasn't that big of a deal.
This actually reminds me, we should make traits for pickle formats too.

eed3si9n commented:
About the self-type annotation, in my head the consensus has not reached on Hangout yet. If I missed that, I would first like to apologize.
When I saw that `DatePicklers` hasn't been converted yet, I thought self-type annotation introduces less constraints. We can discuss further on the [commit](https://github.com/eed3si9n/pickling/commit/0f2b74cca770516c94535bbf36024283eab48367#commitcomment-9366760).

jsuereth commented:
Regarding the PickleFunctions trait

So for you use case #3, I think you need to invert your design (in addition to documenting this intended usage in all the traits, and what they provide).


Specifically:

* Pickler macros and generated code need a *static* method to reference, a.k.a. scala.pickling.functions.<pickle,unpickle,pickleTo> (although I think they only use pickleTo + unpickle).
* If it's implemented using the current mechanism, people may think they can re-implement pickle/unpickle how they choose for their library, as they can all the other traits.  This is not true ("One of these things is not like the other, one of these things just doesn't belong").

I'd rather see this:

```scala
package scala.pickling

object functions {
  def pickle....
  def unpickle...
}

trait PickleFunctions {
   def pickle... = function.pickle...
   def unpickle.... = function unpickle...
}
```

i.e. the PickleFunctions trait is only for a.la. carte (and needs documented as such), while the static functions *are* the implementation, unchanging, in a location the pickler macros can reference.

More importantly, I don't see the PickleFunctions trait (saving two lines of code) really serving its purpose.   Again, if I'm creating my own protocol package, will I even want to expose PickleFormat.OutputType in my API?   For sbt, I'd say *no* we don't.  I don't see us exposing PickleFunctions, and it's dubious to me in what scenario #3 users would want to as well.

jsuereth commented on core/src/main/scala/pickling/pickler/JavaBigDecimal.scala:
```diff
@@ -0,0 +1,34 @@
+/** Contains picklers which serialize java.math.BigDecimal.
+  * Note; This currently serialzies as a string.
+  */
+trait JavaBigDecimalPicklers { self: PrimitivePicklers =>
```
So this trait is pulling in the `stringPickler` as a value, which means it *need* to be instantiated BEFORE this trait.   Which means you need the `is-a` relationship for that to be enforced.

Additionally, the ONLY way to override the default string pickler, for this trait, is via early-initializer.   You can't just use `override val`.


SO, again I ask:   What's the benefit and does it outweigh the cost of users running into runtime NPEs because of trait ordering issues?

Cake is hard, let's not make it harder by using fancy self-types when they're not providing real value.

eed3si9n committed:
Turn functions into object

havocp commented:
Talking to @eed3si9n and @jsuereth now, we had the following discussion:

 * Josh and I feel we should have _either_ allPicklers._ and ops._, _or_ Defaults._, so there's a single path to each implicit and you can't brokenly import the same thing twice and get ambiguous implicits. I like allPicklers._ and ops._ and Josh likes Defaults._. @eed3si9n  likes having all three. @heathermiller / @phaller  could you just tell us what to do here :-) so we can move on.
 * we think we should remove the self type stuff, @eed3si9n is going to update the PR with that
 * @eed3si9n is going to update the PR to invert the PickleFunctions trait and functions object so the trait delegates to the object instead of the object instantiating the trait, so people don't think they can override these functions.

After that we would like to get this merged and publish a release.

Thanks!

havocp commented:
via hangout, @phaller says go with `Defaults._`

phaller commented:
For the record, also agreed on points 2 and 3. So, from my side this will be good to merge afterwards.

eed3si9n committed:
Remove ops and allPicklers in favor of just Defaults

See scala/pickling#267

eed3si9n committed:
Split Ops out of the package object

eed3si9n committed:
Use inheritance instead of self type

jsuereth commented:
Ok, this LGTM.

eed3si9n commented:
I'm writing documentation, and realizing that there's no mixin traits for formats.
Can I make:

```scala
  trait JsonFormats {
    implicit val pickleFormat: JSONPickleFormat = new JSONPickleFormat
    implicit def toJSONPickle(value: String): JSONPickle = JSONPickle(value)
    implicit def jsonPickleToUnpickleOps(value: String): UnpickleOps = new UnpickleOps(JSONPickle(value))   
  }
```

jsuereth commented:
@eed3si9n Does this remove the existing `import scala.pickling.json._` usage or migrate things out of static context?

eed3si9n commented:
@jsuereth No, it'll keep the existing usage intact, but allows someone to make the protocol that forces Pickling Json format.

jsuereth commented:
That seems fine to me.  Does the toJSONPickle/jsonPickletoUnpickleOps stuff exist elsehwere?

eed3si9n committed:
Turn formats into traits

eed3si9n committed:
Avoid ops in macros so the user can mix in ops

eed3si9n commented:
@jsuereth It's currently part of the package object. See the last two commits.

jsuereth commented:
cool, looks good.
